cmd/snap-bootstrap/cmd_scan_disk.go

```diff
@@ -0,0 +1,330 @@
+	hasBoot := false
+	var boot_uuid string
+	hasData := false
+	var data_uuid string
+	hasSave := false
+	var save_uuid string
+	for _, part := range partitions {
+		if !fallback {
+			if part.UUID == bootUUID {
+				found = true
+			}
+		}
+		if fallback && part.Name == fallbackPartition {
+			hasFallbackPartition = true
+		}
+		if part.Name == "ubuntu-seed" {
+			hasSeed = true
+			seed_uuid = part.UUID
+		} else if part.Name == "ubuntu-boot" {
+			hasBoot = true
+			boot_uuid = part.UUID
+		} else if part.Name == "ubuntu-data-enc" {
+			hasData = true
+			data_uuid = part.UUID
+		} else if part.Name == "ubuntu-data" {
+			hasData = true
+			data_uuid = part.UUID
+		} else if part.Name == "ubuntu-save-enc" {
+			hasSave = true
+			save_uuid = part.UUID
+		} else if part.Name == "ubuntu-save" {
+			hasSave = true
+			save_uuid = part.UUID
+		}
+	}
+
+	if (!fallback && found && (hasSeed || hasBoot)) || (fallback && hasFallbackPartition) {
+		fmt.Fprintf(output, "UBUNTU_DISK=1\n")
+		if hasSeed {
+			fmt.Fprintf(os.Stderr, "Detected partition for seed: %s\n", seed_uuid)
+			fmt.Fprintf(output, "UBUNTU_SEED_UUID=%s\n", seed_uuid)
+		}
+		if hasBoot {
+			fmt.Fprintf(os.Stderr, "Detected partition for boot: %s\n", boot_uuid)
+			fmt.Fprintf(output, "UBUNTU_BOOT_UUID=%s\n", boot_uuid)
+		}
+		if hasData {
+			fmt.Fprintf(os.Stderr, "Detected partition for data: %s\n", data_uuid)
+			fmt.Fprintf(output, "UBUNTU_DATA_UUID=%s\n", data_uuid)
+		}
+		if hasSave {
+			fmt.Fprintf(os.Stderr, "Detected partition for save: %s\n", save_uuid)
+			fmt.Fprintf(output, "UBUNTU_SAVE_UUID=%s\n", save_uuid)
+		}
+	}
+
+	return nil
+}
+
+func checkPartitionUUID(output io.Writer, suffix string, partUUID string) {
+	varname := fmt.Sprintf("UBUNTU_%s_UUID", suffix)
+	expectedUUID := osGetenv(varname)
+	if len(expectedUUID) > 0 && expectedUUID == partUUID {
+		fmt.Fprintf(os.Stderr, "Detected partition as %s\n", suffix)
+		fmt.Fprintf(output, "UBUNTU_%s=1\n", suffix)
+	}
+}
+
+func scanPartitionNode(output io.Writer, node string) error {
+	fmt.Fprintf(os.Stderr, "Scanning partition %s\n", node)
+
+	probe, err := blkid.NewProbeFromFilename(node)
+	if err != nil {
+		return err
+	}
+	defer probe.Close()
+
+	probe.EnablePartitions(true)
+	probe.SetPartitionsFlags(blkid.BLKID_PARTS_ENTRY_DETAILS)
+	probe.EnableSuperblocks(true)
+
+	if err := probe.DoSafeprobe(); err != nil {
+		return fmt.Errorf("Cannot probe partition %s: %s\n", node, err)
+	}
+
+	partUUID, err := probe.LookupValue("PART_ENTRY_UUID")
+	if err != nil {
+		return fmt.Errorf("Cannot get uuid for partition: %s\n", err)
+	}
+
+	for _, suffix := range []string{"SEED", "BOOT", "DATA", "SAVE"} {
+		checkPartitionUUID(output, suffix, partUUID)
+	}
+
+	return nil
+}
+
+func ScanDisk(output io.Writer) error {
+	devname := osGetenv("DEVNAME")
+	if osGetenv("DEVTYPE") == "disk" {
+		return scanDiskNode(output, devname)
+	} else if osGetenv("DEVTYPE") == "partition" {
+		return scanPartitionNode(output, devname)
```

Comment:
Do we really need scanning and exporting `UBUNTU_*` if this is a partition? As the env from the (parent) disk is imported, maybe this would work:
```
ENV{DEVTYPE}=="partition", ENV{UBUNTU_DISK}=="1", ENV{PART_ENTRY_NAME}=="ubuntu-seed", SYMLINK+="ubuntu/seed"
...
``` 
and the scan-disk command could be made to run only for the disks and not also for each partition. We would not need to compare uuids.

Reply:
The main point of not matching partition labels in udev rules is because I think we should not do that. And instead match partition type uuids, and register those in https://uapi-group.org/specifications/specs/discoverable_partitions_specification/. And if the logic is just in scan-disk, then there is only one place to care about. It is difficult to make complex logic with backward compliance in udev rules.
Also the "-enc" suffix in labels is really annoying, we never needed that, I hope we can get rid of it. Now the udev rules check the type of the file system to know what to do with it.